src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/HtmlCSharp/OnTypeRenameHandler.cs

```diff
@@ -0,0 +1,111 @@
+        private readonly LSPDocumentMappingProvider _documentMappingProvider;
+        
+        [ImportingConstructor]
+        public OnTypeRenameHandler(
+            LSPDocumentManager documentManager,
+            LSPRequestInvoker requestInvoker,
+            LSPProjectionProvider projectionProvider,
+            LSPDocumentMappingProvider documentMappingProvider)
+        {
+            if (documentManager is null)
+            {
+                throw new ArgumentNullException(nameof(documentManager));
+            }
+
+            if (requestInvoker is null)
+            {
+                throw new ArgumentNullException(nameof(requestInvoker));
+            }
+
+            if (projectionProvider is null)
+            {
+                throw new ArgumentNullException(nameof(projectionProvider));
+            }
+
+            if (documentMappingProvider is null)
+            {
+                throw new ArgumentNullException(nameof(documentMappingProvider));
+            }
+
+            _documentManager = documentManager;
+            _requestInvoker = requestInvoker;
+            _projectionProvider = projectionProvider;
+            _documentMappingProvider = documentMappingProvider;
+        }
+
+        public async Task<DocumentOnTypeRenameResponseItem> HandleRequestAsync(DocumentOnTypeRenameParams request, ClientCapabilities clientCapabilities, CancellationToken cancellationToken)
+        {
+            if (request is null)
+            {
+                throw new ArgumentNullException(nameof(request));
+            }
+
+            cancellationToken.ThrowIfCancellationRequested();
+
+            if (!_documentManager.TryGetDocument(request.TextDocument.Uri, out var documentSnapshot))
+            {
+                return null;
+            }
+
+            var projectionResult = await _projectionProvider.GetProjectionAsync(documentSnapshot, request.Position, cancellationToken).ConfigureAwait(false);
+            if (projectionResult is null || projectionResult.LanguageKind != RazorLanguageKind.Html)
+            {
+                return null;
+            }
+
+            var onTypeRenameParams = new DocumentOnTypeRenameParams()
+            {
+                Position = projectionResult.Position,
+                TextDocument = new TextDocumentIdentifier() { Uri = projectionResult.Uri }
+            };
+
+            var contentType = projectionResult.LanguageKind.ToContainedLanguageContentType();
+            var response = await _requestInvoker.ReinvokeRequestOnServerAsync<DocumentOnTypeRenameParams, DocumentOnTypeRenameResponseItem>(
+                MSLSPMethods.OnTypeRenameName,
+                contentType,
+                onTypeRenameParams,
+                cancellationToken).ConfigureAwait(false);
+
+            if (response is null)
+            {
+                return null;
+            }
+
+            var mappingResult = await _documentMappingProvider.MapToDocumentRangesAsync(
```

Comment:
>var mappingResult = await _documentMappingProvider.MapToDocumentRangesAsync( [](start = 12, length = 76)

Just for my clarity, you always do this, even though the html ranges should be the same, right?

Reply:
I'm pretty sure we do. That being said, I wouldn't be entirely surprised if we flubbed this somewhere along the line.

We're trying to prepare for a day when the HTML mappings don't align 😄